docs/order-with-promo/index.md

```diff
@@ -0,0 +1,627 @@
+
+> Error: A type variable is unbound in this type declaration.
+
+The type variable exists, but it doesn't appear as an argument of the `discount`
+type constructor. Once it's added, it compiles:
+
+<<< Types.re#add-type-arg{1}
+
+This is somewhat like accidentally using a variable in a function but forgetting
+to add that variable to the function's argument list.
+
+## Normal variants vs polymorphic variants
+
+Actually, we are not ready to settle the debate of when to use normal variants
+and when to use polymorphic variants, because polymorphic variants have features
+that we have yet to explore. But in the case of the `discount` reactive value,
+we can say that either would be fine. Still, the version using normal variants
+is a little more explicit and self-documenting, which is never a bad thing.
+
+## Force `DiscountError` argument to be polymorphic variant
+
+Right now the argument of the `DiscountError` constructor can be any type at
+all, but we can force it to be polymorphic constructor:
+
+<<< Types.re#must-be-poly{4}
+
+The `[> ]` type expression means a polymorphic variant that has no tags, but
+allows more tags, which basically means any polymorphic variant.
+
+## Add `DateInput` component
+
+To see different promotions in action, we want to be able to easily change the
+date in our demo, so add a new file `DateInput.re`:
+
+<<< DateInput.re
+
+Note how the `type` prop of `input` has been renamed to `type_`, because in
+OCaml, `type` is a reserved keyword and can't be used as an argument name.
+But don't worry, it will still say `type` in the generated JS output.
+
+## Add `Demo` component
+
+Move the contents of `Index.App` into a new file called `Demo.re`. In the
+process, add our newly-created `DateInput` component:
+
+<<< Demo.re#initial{16-17,21}
+
+Change `Index` to use the new `Demo` component:
+
+<<< Index.re{7}
+
+## Add `Promo` to `Order`
+
+Add the `Promo` component to the `Order` component:
+
+<<< Order.re#make{3,22-25,28}
+
+A breakdown:
+
+- Create a new state variable called `discount`
+- Set the value of `discount` through `Promo`'s `onApply` callback prop (we'll
+  add this prop in the next step)
+- Subtract `discount` from `subtotal` when rendering the total price of the
+  order
+
+## Add `onApply` prop to `Promo`
+
+Add an `onApply` prop to `Promo`, which will be invoked when a promo code is
+successfully submitted and results in a discount:
+
+<<< Promo.re#add-on-apply{3}
+
+::: tip
+
+You don't have to type annotate your component props, but it's a good idea to at
+least type annotate your component's callback props as a form of documentation.
+
+:::
+
+## Add `React.useEffect1`
+
+To invoke `onApply`, we can add a `useEffect` hook that invokes `onApply` when
+`discount` has a value of the form `` `Discount(value) ``:
+
+<<< Promo.re#use-effect
+
+Note that when `discount` has an error value, we return `()` from the switch
+expression, which is essentially a no-op.
```

Comment:
Have we introduced unit before? Otherwise, this is a good chance to point to https://ocaml.org/docs/basic-data-types#unit.